uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/SearchRESTController.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * <p>JSON results example:
+ *
+ * <p>{"people" : [...], "portlets" : [...] }
+ *
+ * @since 5.0
+ */
+@Controller
+@RequestMapping("/search")
+public final class SearchRESTController {
+
+    private static final Logger logger = LoggerFactory.getLogger(SearchRESTController.class);
+
+    @Autowired private ObjectMapper jsonMapper;
+
+    @Autowired private IPersonManager personManager;
+
+    private PersonLookupHelperImpl lookupHelper;
+
+    @Autowired(required = true)
+    public void setPersonLookupHelper(PersonLookupHelperImpl lookupHelper) {
+        this.lookupHelper = lookupHelper;
+    }
+
+    @Resource(name = "directoryQueryAttributes")
+    private List<String> directoryQueryAttributes;
+
+    @Autowired private IPortletDefinitionRegistry portletDefinitionRegistry;
+
+    @Autowired private PortletRegistryUtil portletRegistryUtil;
+
+    @RequestMapping(method = RequestMethod.GET)
+    public void search(
+            @RequestParam("q") String query,
+            HttpServletRequest request,
+            HttpServletResponse response)
+            throws IOException {
+
+        List<Object> matchingPeople = getMatchingPeople(query, request);
+
+        List<Object> matchingPortlets = getMatchingPortlets(query, request);
+
+        if (matchingPeople.isEmpty() && matchingPortlets.isEmpty()) {
+            logger.debug("nothing found");
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+        } else {
+            Map<String, List<Object>> results = new TreeMap<>();
+            results.put("people", matchingPeople);
+            results.put("portlets", matchingPortlets);
+            response.setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
+            jsonMapper.writeValue(response.getOutputStream(), results);
+        }
+    }
+
+    private List<Object> getMatchingPortlets(String query, HttpServletRequest request) {
+        List<Object> results = new ArrayList<>();
+
+        final List<IPortletDefinition> portlets =
+                portletDefinitionRegistry.getAllPortletDefinitions();
+        for (IPortletDefinition portlet : portlets) {
+            if (portletRegistryUtil.matches(query, portlet)) {
+                String url = portletRegistryUtil.buildPortletUrl(request, portlet);
+                if (url != null) {
+                    results.add(getPortletAttrs(portlet, url));
+                }
+            }
+        }
+        return results;
+    }
+
+    private Map<String, String> getPortletAttrs(IPortletDefinition portlet, String url) {
+        Map<String, String> attrs = new TreeMap<>();
+        attrs.put("name", portlet.getName());
+        attrs.put("fname", portlet.getFName());
+        attrs.put("title", portlet.getTitle());
+        attrs.put("description", portlet.getDescription());
+        attrs.put("url", url);
+        return attrs;
+    }
+
+    private List<Object> getMatchingPeople(String query, HttpServletRequest request) {
```

Comment:
No unit test coverage? Test-driven development might drive factoring this kind of thing as something more modular than a `private` method. :smile: